libmamba/include/mamba/core/graph_util.hpp

```diff
@@ -7,38 +7,47 @@
 #ifndef MAMBA_CORE_GRAPH_UTIL_HPP
 #define MAMBA_CORE_GRAPH_UTIL_HPP
 
-#include <map>
 #include <utility>
 #include <vector>
+#include <map>
 
 namespace mamba
 {
     // Simplified implementation of a directed graph
-    // where a path exists between each node and the
-    // first one (you can think of it as a tree with
-    // potential cycles)
-
-    template <class T>
-    class graph
+    template <typename Node, typename GraphCRTP>
+    class DiGraphBase
     {
     public:
-        using node = T;
-        using node_id = size_t;
-        using node_list = std::vector<node>;
-        using edge_list = std::vector<node_id>;
-        using adjacency_list = std::vector<edge_list>;
-        using cycle_list = std::vector<node_list>;
-
-        const node_list& get_node_list() const;
-        const edge_list& get_edge_list(node_id id) const;
-
-        node_id add_node(const node& value);
-        node_id add_node(node&& value);
+        using node_t = Node;
+        using node_id = std::size_t;
+        using node_list = std::vector<node_t>;
+        using node_id_list = std::vector<node_id>;
+        using adjacency_list = std::vector<node_id_list>;
+
+        std::size_t number_of_nodes() const;
+
+        const node_id_list& successors(node_id id) const;
+        const node_id_list& predecessors(node_id id) const;
+
+        const node_list& nodes() const;
+
+        node_id add_node(const node_t& value);
+        node_id add_node(node_t&& value);
         void add_edge(node_id from, node_id to);
 
         template <class V>
         void depth_first_search(V& visitor, node_id start = node_id(0)) const;
 
+    protected:
+        GraphCRTP* crtp_cast()
+        {
+            return static_cast<GraphCRTP*>(this);
+        }
+        GraphCRTP const* crtp_cast() const
+        {
+            return static_cast<GraphCRTP const*>(this);
+        }
```

Comment:
It is usually more convenient to directly return refererences instead of pointers, since most of the APIs expects references. This saves you the indirection.

Nitpicking: I find the name `derived_cast` more explicit than `crtp_cast` and it would be more consistent with the usage of CRTPs we have in other repos that we maintain.